### PR TITLE
[Android]Support external extensions written in C++ on Android

### PR DESCRIPTION
**Pull request by minggangw**
Before this patch, Crosswalk only supports the extensions written in
Java.

With this patch, Crosswalk can load native C++ extensions by calling a
Java API, which will tell the aboslute location to the XWalkExtensionService.
Then when the extension service starts, it will load the extension from
the proper path. To achieve this purpose, add a new option
"--native-extensions" for make_apk.py. By setting it to the location of
the folder, which contains the c++ extension library, the make_apk will
pack them into the final APK file.

Also, this patch creates two unit tests for native external extensions
for runtime client, including shared/embedded modes.

BUG=XWALK-4749

### CODE REVIEW

**Comment by minggangw** [comment]
@halton please review this PR, thanks!

File: xwalk_core_library_android.gypi [hunk 1]
```diff
@@ -26,6 +26,7 @@
           '>(reflection_gen_dir)/wrapper/org/xwalk/core/XWalkJavascriptResult.java',
           '>(reflection_gen_dir)/wrapper/org/xwalk/core/XWalkNavigationHistory.java',
           '>(reflection_gen_dir)/wrapper/org/xwalk/core/XWalkNavigationItem.java',
+          '>(reflection_gen_dir)/wrapper/org/xwalk/core/XWalkNativeExtensionLoader.java',
```
Comment (rakuco) [inline xwalk_core_library_android.gypi]: Please keep this alphabetically sorted.

File: runtime/browser/xwalk_browser_main_parts_android.cc [hunk 2]
```diff
@@ -259,4 +259,12 @@ XWalkExtension* XWalkBrowserMainPartsAndroid::LookupExtension(
   return NULL;
 }
 
+void XWalkBrowserMainPartsAndroid::RegisterExtensionInPath(
+    const std::string& path) {
+  if (extension_service_) {
```
Comment (rakuco) [inline runtime/browser/xwalk_browser_main_parts_android.cc]: Is it actually possible for `extension_service_` to be `NULL` at this point?

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkNativeExtensionLoaderInternal.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2015 The Chromium Authors. All rights reserved.
```
Comment (rakuco) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkNativeExtensionLoaderInternal.java]: Copyright Intel?

File: extensions/common/android/xwalk_native_extension_loader_android.cc [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#include "jni/XWalkNativeExtensionLoaderAndroid_jni.h"
+#include "xwalk/runtime/browser/xwalk_browser_main_parts_android.h"
+#include "xwalk/runtime/browser/xwalk_content_browser_client.h"
+
+namespace xwalk {
+namespace extensions {
+
+void RegisterExtensionInPath(JNIEnv* env, jclass jcaller, jstring path) {
+  const char *str = env->GetStringUTFChars(path, 0);
+  std::string p(str);
```
Comment (rakuco) [inline extensions/common/android/xwalk_native_extension_loader_android.cc]: This doesn't seem to be used anywhere.

File: xwalk_core_library_android.gypi (hunk 1 above)
Comment (minggangw) [inline xwalk_core_library_android.gypi]: Will change the order.

File: runtime/browser/xwalk_browser_main_parts_android.cc (hunk 2 above)
Comment (minggangw) [inline runtime/browser/xwalk_browser_main_parts_android.cc]: Yes, extension_service_ is impossible to be NULL, the 'if' is redundant.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkNativeExtensionLoaderInternal.java (hunk 3 above)
Comment (minggangw) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkNativeExtensionLoaderInternal.java]: I'll change the copyright :)

File: extensions/common/android/xwalk_native_extension_loader_android.cc (hunk 4 above)
Comment (minggangw) [inline extensions/common/android/xwalk_native_extension_loader_android.cc]: This local variable should be removed here, thank you!

**Comment by rakuco** [comment]
Thanks, the code looks better now. I still have a few comments/ideas:
* Your actual commit message is very different from the pull request message. The former only says "XWALK-4749" without anything else. You should use the latter instead.
* You should probably make the help message for the new make_apk option more informative: it does not say that the extensions need to be in architecture-specific directories inside the directory being passed, and their names are not mentioned anywhere. Adding some more information like "Found native extension of architecture XYZ" when doing the processing is also helpful.
* The lack of tests makes it prone to regressions in the future. You should at least add a test native extension to verify everything is working as expected.

**Comment by minggangw** [comment]
Thank you for your suggestions, I will add some tests to verify this new feature.

**Comment by minggangw** [comment]
Hi, @rakuco . I did some investigations on how to write an automation test to verify the native extension on Android (I supposed the test you mentioned is an automation test) and found some problems:

1. In order to test the native extension, we have to make a real apk file and run it on emulator/device. I think to achieve this we have a lot of things to do.
2. The native C++ extension mechanism itself have been tested on other platform like the  "echo_extension". Here, we just reuse it. The function doesn't need to be tested again.

I discussed with Halton about it and have a sample solution. We can reuse the current native extension C++ test case, write the detailed description on how to test it manually in the commit log. So what is your opinion, thanks!

**Comment by rakuco** [comment]
> In order to test the native extension, we have to make a real apk file and run it on emulator/device. I think to achieve this we have a lot of things to do.

The whole setup to build and run device tests is already in place, and that's how we run all the tests [here](https://build.crosswalk-project.org/builders/Crosswalk%20Tests%20%28Android%20x86%29), for example. IIRC we even test Java extensions this way already. Isn't it just a matter of building another simple extension the way and extending what we already do?

> The native C++ extension mechanism itself have been tested on other platform like the "echo_extension". Here, we just reuse it. The function doesn't need to be tested again.
>
> I discussed with Halton about it and have a sample solution. We can reuse the current native extension C++ test case, write the detailed description on how to test it manually in the commit log. So what is your opinion, thanks!

The problem with this idea is that it basically assumes that:
- The code will not substantially change in the future, and thus the way the C++ extensions work and how they are plugged into the Android code will always work.
- People will somehow know that in this specific commit from a random date in the project's history there are instructions on how to test the C++/Java integration of the extensions in the commit message. Even if the instructions are in a comment in the code instead, John Doe who starts contributing to the project 2 years from now will have no idea that the instructions are there.

Which is why it would be nice to do these things automatically: you just run your usual test suite and know if your changes broke something or not.

IIRC @wuhengzhi added several Android test cases in the past, so he might be able to help you with this. If you still have problems, I can also try to give you a hand.

**Comment by minggangw** [comment]
Ok, I will contact @wuhengzhi .

**Comment by minggangw** [comment]
@rakuco , I've updated the patch by your comments, adding 2 tests and some helpful informations. The instrumentation tests have passed by running test_runner.py instrumentation --release --test-apk=XWalkRuntimeClientEmbeddedTest/XWalkRuntimeClientTest. Please have a look, thank you. And thanks for @wuhengzhi 's help :)

**Comment by minggangw** [comment]
Don't know why the PR closed, just reopen it.

**Comment by halton** [comment]
ping @rakuco

**Comment by halton** [comment]
@rakuco how is your review status on this PR?

**Comment by halton** [comment]
@wuhengzhi , you're right. @minggangw please fix the style thing.

**Comment by minggangw** [comment]
Ok, I will fix that.

**Comment by halton** [comment]
@rakuco the style issues are fixed, please continue your review.

**Comment by rakuco** [comment]
Sorry, I'm focused on M45 at the moment and don't have much time to context switch back to this PR. Overall it looks fine to me now; if it's OK to you and @wuhengzhi then feel free to go ahead.

File: xwalk_android_tests.gypi [hunk 5]
```diff
@@ -26,6 +26,17 @@
       'includes': [ '../build/java.gypi' ],
     },
     {
+      'target_name': 'native_echo_extension',
+      'type': 'loadable_module',
+      'include_dirs': [
+        '..',
+      ],
+      'sources': [
+        'test/android/data/nativeechoextension/native_echo_extension.c',
```
Comment (halton) [inline xwalk_android_tests.gypi]: I presume the content of native_echo_extension.c is same as `src/xwalk/extensions/test/echo_extension.c`. Can we re-use the the code and target instead of copy?

File: xwalk_android_tests.gypi [hunk 6]
```diff
@@ -248,12 +259,18 @@
       'dependencies': [
         'xwalk_app_runtime_java',
         'xwalk_runtime_client_test_utils_java',
+        'libxwalkdummy',
```
Comment (halton) [inline xwalk_android_tests.gypi]: I do not think this change is related. If you're trying to fix some bug, please submit separate PR

File: xwalk_android_tests.gypi [hunk 7]
```diff
@@ -440,6 +459,7 @@
           '<(PRODUCT_DIR)/runtime_client_test/assets/echo.html',
           '<(PRODUCT_DIR)/runtime_client_test/assets/echoSync.html',
           '<(PRODUCT_DIR)/runtime_client_test/assets/messaging_mini.html',
+          '<(PRODUCT_DIR)/runtime_client_embedded_test/assets/nativeEcho.html',
```
Comment (halton) [inline xwalk_android_tests.gypi]: ditto, please re-use the echo.html instead of copy.

File: xwalk_android_tests.gypi (hunk 5 above)
Comment (minggangw) [inline xwalk_android_tests.gypi]: In the current RuntimeClinetShell/RuntimeClientEmbeddedShell test, there is already a Java external extension named "echo" as the js interface, see "test/android/data/extensions-config.json". 
"/test/echo_extension.c" defined the same "echo" as the extension name. So, here I redefine "nativeEcho" as the name of the extension to distinguish the one written in Java.  Most of the codes in echo_extension.c and native_echo_extension.c are the same, excepted the name of the js interface.

File: xwalk_android_tests.gypi (hunk 6 above)
Comment (minggangw) [inline xwalk_android_tests.gypi]: I will check if 'libxwalkdummy' is necessary here, thanks!

**Comment by minggangw** [comment]
After discussing with @halton about the conflicts of the existing naming "echo" js interface in ExternalExtensionTest/NativeExternalExtensionTest tests (one for Java external extension, the other is native c++ extension), I prefer to solve it:

1. Rename the current test case resources (htmls, js interface) of Android Java to something like, xxx_java to indicate that this is a tc for Java extension. Then commit it as the first patch.
2. Commit the test case for native c++ extension as the second patch, which will reuse the current test/echo_extension.c and its test case.

By doing this, we can reduce repetitive test code of c++ and just rename the current Java test case to avoid conflicting.

**Comment by minggangw** [comment]
Follow my previous comment, I submitted the fist patch to rename the js interface. To ensure these changes of renaming don't break the original instrumentation tests, XWalkCoreInternalTest/XWalkCoreTest/XWalkRuntimeClientTest/XWalkRuntimeEmbeddedClinetTest were tested. Please @halton have a look, thanks!

**Comment by wuhengzhi** [comment]
In the commit log, ```echoSync_jave``` -> ```echoSync_java```, ```chaged``` -> ```changed```.

**Comment by halton** [comment]
@minggangw you should open a new PR for the renaming thing and leave this PR for the feature you worked on. Because all the discussion of this PR is about your feature.

**Comment by minggangw** [comment]
Sorry for the wrong spelling, I will open a new PR to do the naming issue.

File: xwalk_android_tests.gypi [hunk 8]
```diff
@@ -248,12 +248,17 @@
       'dependencies': [
         'xwalk_app_runtime_java',
         'xwalk_runtime_client_test_utils_java',
+        'extensions/external_extension_sample.gyp:echo_extension',
```
Comment (halton) [inline xwalk_android_tests.gypi]: nit, order plz.

File: app/tools/android/make_apk.py [hunk 9]
```diff
@@ -757,6 +776,13 @@ def main(argv):
           'Mac OS respectively. For example, '
           '--extensions=/path/to/extension1:/path/to/extension2.')
   group.add_option('--extensions', help=info)
+  info = ('The native external extension path. Put the architecture-specific '
+          'library file in the corresponding folder (armeabi-v7a/x86) under '
```
Comment (halton) [inline app/tools/android/make_apk.py]: There are more archs: x86-64, arm64 to be supported. I think you need verify whether the .so arch is matched with current app_template.

**Comment by minggangw** [comment]
@halton PR updated, pls have a look, thanks!

File: app/tools/android/make_apk.py [hunk 10]
```diff
@@ -95,6 +106,21 @@ def GetAndroidApiLevel(android_path):
...
+  # the length is 2 bytes.
+  header = library.read(20)
+  return struct.unpack_from('H', header, 18)[0]
+
+def CheckValidationOfExpectedLibraryArch(path, expected_arch):
+  extension_path_list = FindNativeExtensionFiles(path);
+  for extension_path in extension_path_list:
+    if not ConvertArchNameToArchValue(expected_arch) == \
+        GetNativeLibraryArchValueInELFHeader(extension_path):
+      print ('Invalid CPU arch of %s, %s is expected.' % (extension_path,
```
Comment (halton) [inline app/tools/android/make_apk.py]: nit, remove space after `print`

Comment (minggangw) [inline app/tools/android/make_apk.py]: Ok

Comment (halton) [inline app/tools/android/make_apk.py]: Other part LGTM.

**Comment by halton** [comment]
LGTM, merging.
